functools/README.md

```diff
@@ -0,0 +1,63 @@
+# functools
+
+## Prerequisites
+
+* [winglang](https://winglang.io).
+
+## Installation
+
+```sh
+npm i @winglibs/functools
+```
+
+## Usage
+
+### `IdempotentFunction`
+
+The `IdempotentFunction` class, inspired by [Powertools for AWS
+Lambda](https://docs.powertools.aws.dev/lambda/typescript/latest/utilities/idempotency/),
+provides a simple solution to convert your inflight functions into idempotent
+operations which are safe to retry. Under the hood, it extends `cloud.Function` to communicate with a persistence layer to prevent handlers from executing
+more than once on the same event payload, and it ensures handlers returns the
+same result when called with the same payload.
+
+```js
+bring functools;
+
+let submitOrder = new functools.IdempotentFunction(inflight (order: Json) => {
```

Comment:
It is possible to design this API so it can be composed with other `cloud.Function` "extensions" instead of subclass?

Can we get to something like this?

```js
let f = new cloud.Function(...);

funtools.addIdempotency(f);
```

Reply:
> Can we get to something like this?

It's possible, though it would require some form of inversion of control or middleware mechanism, eg

```js
let f = new cloud.Function(...);
f.useMiddleware(inflight (handler) => {
  return inflight (payload) => {
    // insert wrapper logic...
    let result = handler(payload);
    return result;
  };
});
funtools.addIdempotency(f);
```

But it's a bit restrictive because the inflight handler would still need to be `inflight (str?): str?` at the end of the day

Reply:
So how would you recommend we compose multiple functools?

Reply:
I guess it depends, what kinds of other functools do you have in mind ? :) addMetrics? addTelemetry? addCircuitBreaker?

The middleware mechanism could probably work alright - if you want to take a stab at implementing something, I'd say go for it 👍  it seems pretty powerful.

Also `funtools` could be a nice name...